rfcs/remote-schema-permissions.md

```diff
@@ -0,0 +1,244 @@
+
+Each field in GraphQL can take arguments as well. Arguments must belong to input types i.e Scalar, Enum, or Input Object (or list of these).
+
+Role-based argument presets can basically inject values from session variables or static values during execution. The presetted argument should also get removed from the exposed role-based schema.
+
+The proposed API:
+
+```
+type: add_remote_schema_permissions
+args:
+  remote_schema: comm
+  role: user
+  definition:
+    allowed_objects:
+    - name: Query
+      fields: 
+      - name: hello
+      - name: user
+        argument_presets:
+        - name: "id"
+          value:
+            from_session_variable: "x-hasura-user-id"
+    - name: User
+      fields:
+      - name: a
+      - name: b
+      - name: c
+```
+
+Argument presets can also go arbitrarily deep if the input type is Input Object (this is very similar to how we define the mapping in remote relationships):
+
+```
+type: add_remote_schema_permissions
+args:
+  remote_schema: comm
+  role: user
+  definition:
+    allowed_objects:
+    - name: Query
+      fields:
+        - name: hello
+        - name: user
+          argument_presets:
+          - name: "where"
+            input_field:
+            - name: "id"
+              input_field:
+              - name: "_eq"
+                value:
+                  from_session_variable: "x-hasura-user-id"
+          - name: "limit"
+            value:
+              static: 1
+    - name: User
+      fields:
+      - name: a
+      - name: b
+      - name: c
+```
+
+There can be more than one argument preset per field:
+
+```
+type: add_remote_schema_permissions
+args:
+  remote_schema: comm
+  role: user
+  definition:
+    allowed_objects:
+    - name: Query
+      fields: 
+      - name: hello
+      - name: user
+        argument_presets:
+        - name: "id"
+          value:
+            from_session_variable: "x-hasura-user-id"
+        - name: "limit"
+          value:
+            static: 1
+    - name: User
+      fields:
+      - name: a
+      - name: b
+      - name: c
+```
+
+In the above example, if the client (as role "user") makes the query:
+
+```
+query {
+  user {
+    a
+    b
+  }
+}
+```
+
+then the query that is constructed for the remote schema is
+
+```
+query {
+  user(id: "<x-hasura-user-id>", limit: 1) {
+    a
+    b
+  }
+}
+```
+
+## Few questions to tackle:
+
+1. This will be a breaking change as all remote schemas will be hidden for all role except `admin`.
+This was tackled in #2690 by introducing a feature flag for enabling remote schema permissions explicitly.
+2. UX of this feature in the console.
+3. The API proposed above is complete in defining the permission for a role and a remote schema. Should we have smaller additive permission API e.g. 
+one which takes permissions for a single type (of any type) and adds them together to yield the final schema?
```

Comment:
I'm also wondering what happens when the schema changes; seems like the whitelisting approach might be difficult to maintain? On the other hand it's a way of getting a more stable or simple schema even when the remote changes.

Reply:
Good question. Will add it to the list.